### ISSUE TITLE
NIC queue workload monitor tool suggestion

### ISSUE THREAD

**Issue by Yolandajn** [iovisor/bcc#3037]
I would like to contribute a new tool netqtop for monitoring traffic loads on each queue of a specified network interface. The NICs often have multiple queues for both transmitting and receiving data, and it is better to have the workloads balanced on each queue to reach the best performance. However sometimes flow hashing algorithm does not work as expected or other problems cause imbalance. So here comes the request of this tool to help us know if the workloads are balanced or not.

Measurements of workload are PPS, BPS, average size and total number of packets of different size ranges.  And netqtop tool is to print these statistics every interval seconds (default 1).

For example, 

    Tue Aug  4 15:16:41 2020
    TX
     QueueID      BPS        PPS      avg_size   [0, 64)   [64, 512)  [512, 2K)  [2K, 16K)  [16K, 64K)
        0         xxx        xxx        xxx       xxx       xxx        xxx        xxx        xxx
        1         xxx        xxx        xxx       xxx       xxx        xxx        xxx        xxx
     Total        xxx        xxx        xxx       xxx       xxx        xxx        xxx        xxx

    RX
     QueueID      BPS        PPS      avg_size   [0, 64)   [64, 512)  [512, 2K)  [2K, 16K)  [16K, 64K)
        0         xxx        xxx        xxx       xxx       xxx        xxx        xxx        xxx
        1         xxx        xxx        xxx       xxx       xxx        xxx        xxx        xxx
     Total        xxx        xxx        xxx       xxx       xxx        xxx        xxx        xxx

I have developed a program that accomplish these above, using net tracepoints and bpf maps.

**Comment by yonghong-song** [comment]
Some nic drivers exposed such statistics through `ethtool -S`. For example, on my system, for broadcom nic,
```
-bash-4.4$ ethtool -S eth0
NIC statistics:
     [0]: rx_ucast_packets: 2452705234
     [0]: rx_mcast_packets: 26947693
     [0]: rx_bcast_packets: 13655
     [0]: rx_discards: 0
     [0]: rx_drops: 284671450
     [0]: rx_ucast_bytes: 2719544414454
     [0]: rx_mcast_bytes: 5079255428
     [0]: rx_bcast_bytes: 852604
     [0]: tx_ucast_packets: 4333264471
     [0]: tx_mcast_packets: 448793
     [0]: tx_bcast_packets: 0
     [0]: tx_discards: 0
     [0]: tx_drops: 0
     [0]: tx_ucast_bytes: 5651194422855
     [0]: tx_mcast_bytes: 88208648
     [0]: tx_bcast_bytes: 0
     [0]: tpa_packets: 1761542148
     [0]: tpa_bytes: 2472433210833
     [0]: tpa_events: 108993561
     [0]: tpa_aborts: 329621451
     [0]: rx_l4_csum_errors: 258
     [0]: rx_buf_errors: 0
     [0]: missed_irqs: 0
     [1]: rx_ucast_packets: 2385126830
     [1]: rx_mcast_packets: 18806211
     [1]: rx_bcast_packets: 10
     [1]: rx_discards: 0
     [1]: rx_drops: 0
     [1]: rx_ucast_bytes: 2655153297122
...
```
What kinds of drivers in your machine? I understand the output might be more useful than `ethtool` output but user can easily construct their own script to do a post-processing of the raw data. Any reason the data itself collected by bpf program more interesting than nic statistics?

**Comment by Yolandajn** [comment]
yes bpf program privides statistics that nic driver does not. Here are my reasons for using this tool other than ethtool:
1. It is possible to calculate PPS or BPS from ethtool raw data but not number of packets in given size range. For example, netqtop provides packet counts of range [0, 64)   [64, 512)  [512, 2K)  [2K, 16K)  [16K, 64K) which is not provided by ethtool. And sometimes we are interested about the distribution of packet size. 
2. ethtool only works with drivers that supports it. For those drivers they do not support ethtool, there's no data exposed.
```
[root@x ~]# ethtool -S lo
no stats available
```
or no enough data:
```
[root@x ~]# ethtool -S bound0
no stats available
```
3. As the raw data is provided by nic that supports ethtool, the formats differ.  So we have to write different script for each unique driver. The tool [ethq](https://github.com/isc-projects/ethq) has done this job. It supports 7 groups of drivers (still a small number), implementing different parser for each of them respectively. Unfortunately we have a lot of NICs with different drivers to be tested. It would be torturous to write a script for every nic. Hence I use netqtop: it can collect data for all NICs in Linux computers regardless of the drivers.

**Comment by yonghong-song** [comment]
Yes, I agree that nic driver does not have packet size distributions and also one tool for all drivers sound nice esp. not all drivers have all stats. The only concern is potential high overhead since the event is triggered by every packet. I think we can put some warning/explanation in the tool to mention such potential overhead so users will be aware. Overall, I think the tool like this should be useful. So Thanks, please go ahead to submit a pull request. FYI. The following is the link about how to write a bcc tool.
https://github.com/iovisor/bcc/blob/master/CONTRIBUTING-SCRIPTS.md
